streaming/base/dataset/streaming.py

```diff
@@ -0,0 +1,639 @@
+                shared memory.
+            shard_id (int): Shard ID.
+        """
+        # First, the fast path: check the shared memory shard state without taking the lock. The
+        # shard states only ever go up, so if we're at the downloaded state, it's downloaded.
+        state = shard_states[shard_id]
+        if state == _ShardState.DOWNLOADED:
+            return
+
+        # Shard is not necessarily downloaded, so check and update state with the lock.
+        lock.acquire()
+        state = shard_states[shard_id]
+        if state == _ShardState.UNKNOWN:
+            shard_states[shard_id] = _ShardState.DOWNLOADING
+            lock.release()
+            self._download_shard(shard_id)
+            with lock:
+                shard_states[shard_id] = _ShardState.DOWNLOADED
+        elif state == _ShardState.DOWNLOADING:
+            lock.release()
+            while shard_states[shard_id] != _ShardState.DOWNLOADED:
+                sleep(0.07)
+        elif state == _ShardState.DOWNLOADED:
+            lock.release()
+        else:
+            raise RuntimeError('Unknown shard state')
+
+    def _download_thread(self, epoch: int, sample_ids: NDArray[np.int64]) -> None:
+        """Download the relevant shards in the background while we are being iterated.
+
+        Args:
+            epoch (int): Which epoch. On noticing that a new epoch has started, we exit this thread
+                because a new one will soon be running, with different sample IDs.
+            sample_ids (NDArray[np.int64]): The samples to download the shards of.
+        """
+        # Create or attach shard_states array.
+        name = f'{self._prefix}_shard_states'
+        size = len(self.shard_sizes) * np.uint8(0).nbytes
+        try:
+            shm = SharedMemory(name, True, size)
+        except:
+            shm = SharedMemory(name)
+        shard_states = np.ndarray(len(self.shard_sizes), buffer=shm.buf, dtype=np.uint8)
+
+        filename = os.path.join('/tmp', 'mds', self._prefix, '_shard_states_filelock')
+        shard_states_lock = FileLock(filename)
+
+        # Download loop.
+        num_samples = len(sample_ids)
+        while True:
+            # If we've started a new epoch early (__iter__ was called again), exit this thread
+            # because there can only be one epoch at once.
+            if epoch != self.next_epoch - 1:
+                break
+
+            # If we're out of samples this epoch, exit this thread because we are done downloading.
+            if self._download_index == num_samples:
+                break
+
+            # If we are requested to only pre-download so many samples, if we have as many or more
+            # downloaded already, we wait and check again later.
+            if self.predownload is not None:
+                samples_ahead = self._download_index - self._iter_index
+                if self.predownload <= samples_ahead:
+                    sleep(0.07)
```

Comment:
are you a big fan of James Bond movie ? ;) wondering where does `0.07` value comes from ?

Reply:
Please assign all the constant at the top of the file with ALL_CAPS.

Reply:
kNiGhToN, jAmEs KnIgHtOn

Reply:
done